src/canisters/telemetry/src/telemetry_backend/main.mo

```diff
@@ -0,0 +1,208 @@
+import Debug "mo:base/Debug";
+import Hash "mo:base/Hash";
+import Int "mo:base/Int";
+import Nat "mo:base/Nat";
+import Option "mo:base/Option";
+import Prelude "mo:base/Prelude";
+import Time "mo:base/Time";
+import Trie "mo:base/Trie";
+
+import Aggregation "Aggregation";
+import CommandDetails "CommandDetails";
+import CommandResults "CommandResults";
+import Data "Data";
+import InvocationDetails "InvocationDetails";
+import Interface "Interface";
+
+actor {
+  type Time = Time.Time;
+  type Trie<K, V> = Trie.Trie<K, V>;
+  type Trie2D<K1, K2, V> = Trie.Trie2D<K1, K2, V>;
+
+  type AggregationPeriodStart = Aggregation.AggregationPeriodStart;
+  type CommandDetails = CommandDetails.CommandDetails;
+  type InvocationDetails = InvocationDetails.InvocationDetails;
+  type CommandResultsEntry = Interface.CommandResultsEntry;
+  type CommandResultCounts = CommandResults.CommandResultCounts;
+  type CommandSuccessRatesEntry = Interface.CommandSuccessRatesEntry;
+  type ReportCommandArgs = Interface.ReportCommandArgs;
+  type V0 = Data.V0;
+
+  let nsPerDay = 86_400 * 1000_000_000;
+  let nsPer30Days = 30 * nsPerDay;
+
+  stable var v0 : ?V0 = ?Data.v0();
+  let data : V0 = switch (v0) {
+    case null Prelude.unreachable();
+    case (?v0) v0;
+  };
+
+  func currentTime() : Time {
+    switch (data.overrideTime) {
+      case null Time.now();
+      case (?override) override;
+    }
+  };
+
+  func updateAggregationPeriods() {
+    let now = currentTime();
+
+    if (now > data.dailyAggregationPeriodEnd) {
+      let day : Time = now / nsPerDay;
+      data.dailyAggregationPeriodStart := day * nsPerDay;
+      data.dailyAggregationPeriodEnd :=
+        data.dailyAggregationPeriodStart + nsPerDay - 1;
+    };
+
+    if (now > data.thirtyDayAggregationPeriodEnd) {
+      if (data.thirtyDayAggregationPeriodStart == 0) {
+        data.thirtyDayAggregationPeriodStart :=
+          data.dailyAggregationPeriodStart;
+      } else {
+        let periodsToAdd =
+          (now - data.thirtyDayAggregationPeriodStart) / nsPer30Days;
+        data.thirtyDayAggregationPeriodStart += periodsToAdd * nsPer30Days;
+      };
+      data.thirtyDayAggregationPeriodEnd :=
+        data.thirtyDayAggregationPeriodStart + nsPer30Days - 1;
+    };
+  };
+
+  public func reportCommandResult(args: ReportCommandArgs) : async () {
```

Comment:
I understand ACL is a separate ticket, but how do we plan to prevent non-dfx client sending data to this endpoint?

Reply:
We aren't going to do anything to prevent it.  It's a public update method on the canister.  For the first iteration, dfx is probably going to actually spawn  `dfx canister call --network ic --identity anonymous ...`, since it's reporting success/failure of the command, in order not to slow down developer flow while waiting for the call to go through.

The only real measure taken in this regard is to store enums rather than strings (`#dfxCanisterCall` vs `"dfx canister call"`).  The only Text value stored is the dfx version, and we can at least verify that it's some subset of semver.

Reply:
Sounds like you want to make this a one-way function, i.e. change the return type from `async ()` to `()`. This way canister call will return immediately and don't care about whether the update call succeed or not.

My concern with ACL is that one malicious user, e.g., calling this function all the time, can skew the whole aggregate result. But I don't have a good solution either.

Reply:
Thanks!  I didn't know about that distinction.

The references to ACL in the PR description are about controlling access to the methods used for testing (testSetTime and testGetCommandResults).  As for the rest, I agree about the potential problem and the lack of a good solution.